[GH-15025] Soft delete user DM channels when user is deactivated

barisaydogdu opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
This PR ensures that when a user is deactivated, all their direct message (DM) channels are soft-deleted by setting DeleteAt timestamps. This aligns DM channel behavior with deactivation logic for other channels and prevents orphaned active channels for deactivated users.

QA can verify this by:

Creating multiple DM channels with a user.

Deactivating the user account.

Verifying that DeleteAt is properly set for all associated DM channels.


#### Ticket Link

Fixes https://github.com/mattermost/mattermost/issues/15025


#### Screenshots
N/A – No UI changes.

#### Release Note

Soft delete all user direct message channels upon user deactivation to maintain consistent data handling.


Examples:

```release-note
NONE
```

mattermost-build commented:
Hello @barisaydogdu,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

Per the [Mattermost Contribution Guide](https://developers.mattermost.com/contribute/getting-started/), we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**
Once you have signed the CLA, please comment with `/check-cla` and confirm that the CLA check is green.

This is a standard procedure for many open source projects.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

barisaydogdu commented:
/check-cla

barisaydogdu commented:
> > Hi team, since this PR comes from a fork, workflows are pending approval and at least one review is required. Could someone please approve the workflows and review the PR when you have a chance? Thanks!
> 
> I've approved the workflows and added initial reviewers. Looks like there are some failing tests to resolve here.

@amyblais 
I’ve fixed the failing tests. Could someone with the required permissions please approve the workflows for this PR again so that the tests can proceed? The review has already been requested from the reviewers, but the workflows are currently pending approval. Thanks a lot!

amyblais commented:
> I’ve fixed the failing tests. Could someone with the required permissions please approve the workflows for this PR again so that the tests can proceed? The review has already been requested from the reviewers, but the workflows are currently pending approval. Thanks a lot!

For sure, I've approved the workflows for this PR. There may still be some failing tests.

Willyfrog commented:
/update-branch

barisaydogdu commented:
@Willyfrog 

Hi,
At the time I pushed my code, these tests didn’t exist yet. Later, I fixed a small issue related to my own changes, and all tests were passing locally. The failing tests now seem to be the ones that were added afterwards. I also pulled the latest changes into my fork, and it looks like since the PR is open, these new tests are now running together with my changes, which is why they are failing. I might be wrong, but from what I can see, these errors are not caused by the code I wrote. I’ll pull the most recent changes and push again shortly.

barisaydogdu commented:
@Willyfrog @amyblais 

I’ve updated my branch with the latest changes and re-ran all tests locally — everything is passing now without any issues.
The previous failures were caused by new tests that didn’t exist when I initially pushed my code, so they weren’t related to my changes.
Could someone with the required permissions please approve the workflows for this PR again so the tests can proceed?
Thanks a lot!

amyblais commented:
> The linter started failing on my PR due to the new naming convention for request.CTX parameters:
> 
> channels/app/user.go:1048:40: parameter of type request.CTX should be named 'rctx', got 'c'
> 
> It looks like the convention for request.CTX parameters has changed recently. I’m updating the affected parameter names (e.g. from c to rctx) to fix the linter errors

Thank you!

barisaydogdu commented:
> I'm pretty sure it has nothing to do with your changes. I'll ask someone to take a look

Thank you!

Willyfrog commented:
Once this PR is in, you'll have to rebase

https://github.com/mattermost/mattermost/pull/33920

sorry for the inconvenience! 🙏🏼

isacikgoz commented on server/i18n/en.json:
Probably a formatting error here, can you please take a look at it? Thanks!

isacikgoz commented on server/channels/app/user.go:
```diff
@@ -1038,6 +1038,58 @@ func (a *App) userDeactivated(rctx request.CTX, userID string) *model.AppError {
+	return nil
+}
+
+func (a *App) softDeleteUserDMChannels(rctx request.CTX, userID string, deleteTimestamp int64) *model.AppError {
```
One potential issue here is, if the DM channel has more than 2 members (ie. a group DM channel), this will also automagically deleting that channel too. Just want to confirm this from the UX point of view to understand best approach. cc @matthewbirtch @abhijit-singh

isacikgoz commented on server/channels/app/user.go:
```diff
@@ -1038,6 +1038,58 @@ func (a *App) userDeactivated(rctx request.CTX, userID string) *model.AppError {
+				continue
+			}
+
+			if err := a.Srv().Store().Channel().SetDeleteAt(channel.Id, deleteTimestamp, deleteTimestamp); err != nil {
```
Why not use `App.DeleteChannel` method? 

https://github.com/mattermost/mattermost-server/blob/d78d59babeb994106e305531de50b8d515427396/server/channels/app/channel.go#L1498

isacikgoz commented on server/channels/app/user.go:
```diff
@@ -1038,6 +1038,58 @@ func (a *App) userDeactivated(rctx request.CTX, userID string) *model.AppError {
+			if channel.Type != model.ChannelTypeDirect {
+				continue
+			}
+			if channel.DeleteAt != 0 {
+				continue
+			}
```
Is it necessary still if you already queried with `includeDeleted` as `false`?

isacikgoz commented on server/channels/app/user.go:
```diff
@@ -1038,6 +1038,58 @@ func (a *App) userDeactivated(rctx request.CTX, userID string) *model.AppError {
+		}
+		if len(channels) == 0 {
+			break
+		}
```
This is redundant as you are already checking the size of queried channels below: https://github.com/mattermost/mattermost/pull/33774/files#diff-03e2702288b2a1348d400efa6e5ac2dfc6748997f099f46a2366bc8a85efa357R1088

isacikgoz reviewed (changes requested):
Thanks for the contribution @barisaydogdu 🚀 I have a few comments to discuss and also some change requests.

isacikgoz reviewed (commented):
LGTM, just need to fix `i18n.json` file and one small comment.

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!
